### PR TITLE
🎉 New Connector: Clockify [python cdk]

### PR DESCRIPTION
**Pull request by RobLucchi**
## What
New Source Connector For [Clockify](https://clockify.me/) 

## How
This is a new Source Connector for [Clockify](https://clockify.me/)  using [Python CDK](https://docs.airbyte.com/connector-development/cdk-python/)

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [x] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret`
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [x] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

</details>

## Tests
<details><summary><strong>Unit Tests</strong></summary>

airbyte-integrations/connectors/source-clockify/unit_tests/test_source.py::test_check_connection ✓                                                                                                            14% █▌        
 airbyte-integrations/connectors/source-clockify/unit_tests/test_source.py::test_streams ✓                                                                                                                     29% ██▉       
 airbyte-integrations/connectors/source-clockify/unit_tests/test_streams.py::test_request_params ✓                                                                                                             43% ████▍     
 airbyte-integrations/connectors/source-clockify/unit_tests/test_streams.py::test_next_page_token ✓                                                                                                            57% █████▊    
 airbyte-integrations/connectors/source-clockify/unit_tests/test_streams.py::test_read_records ✓                                                                                                               71% ███████▎  
 airbyte-integrations/connectors/source-clockify/unit_tests/test_streams.py::test_request_headers ✓                                                                                                            86% ████████▋ 
 airbyte-integrations/connectors/source-clockify/unit_tests/test_streams.py::test_http_method ✓                                                                                                               100% ██████████

Results (0.24s):
       7 passed

</details>

<details><summary><strong>Integration Tests</strong></summary>

Test session starts (platform: linux, Python 3.10.6, pytest 6.2.5, pytest-sugar 0.9.5)
cachedir: .pytest_cache
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/root/Development/airbyte/airbyte-integrations/connectors/source-clockify/.hypothesis/examples')
rootdir: /root/Development/airbyte, configfile: pytest.ini
plugins: requests-mock-1.9.3, mock-3.6.1, sugar-0.9.5, cov-3.0.0, timeout-1.4.2, hypothesis-6.54.6
collecting ... 
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_config_match_spec[inputs0] ✓                      4% ▍         
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_match_expected[inputs0] ✓                         8% ▊         
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_docker_env[inputs0] ✓                            12% █▎        
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_oneof_usage[inputs0] ✓                           15% █▋        
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_required[inputs0] ✓                              19% █▉        
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_optional[inputs0] ✓                              23% ██▍       
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_has_secret[inputs0] ✓                            27% ██▊       
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_secret_never_in_the_output[inputs0] ✓            31% ███▏      
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_defined_refs_exist_in_json_spec_file[inputs0] ✓  35% ███▌      
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_oauth_flow_parameters[inputs0] ✓                 38% ███▉      Pulling docker image airbyte/source-clockify:latest
{"type": "LOG", "log": {"level": "WARN", "message": "\n We did not find the airbyte/source-clockify:latest image for this connector. This probably means this version has not yet been published to an accessible docker registry like DockerHub."}}

 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_backward_compatibility[inputs0] s                42% ████▎     
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_additional_properties_is_true[inputs0] ✓         46% ████▋     
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestConnection.test_check[inputs0] ✓                           50% █████     
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestConnection.test_check[inputs1] ✓                           54% █████▍    
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestDiscovery.test_discover[inputs0] ✓                         58% █████▊    
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestDiscovery.test_defined_cursors_exist_in_schema[inputs0] ✓  62% ██████▎   
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestDiscovery.test_defined_refs_exist_in_schema[inputs0] ✓     65% ██████▋   
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestDiscovery.test_defined_keyword_exist_in_schema[inputs0-allOf] ✓69% ██████▉   
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestDiscovery.test_defined_keyword_exist_in_schema[inputs0-not] ✓73% ███████▍  
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestDiscovery.test_primary_keys_exist_in_schema[inputs0] ✓     77% ███████▊  
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestDiscovery.test_streams_has_sync_modes[inputs0] ✓           81% ████████▏ 
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestDiscovery.test_additional_properties_is_true[inputs0] ✓    85% ████████▌ Pulling docker image airbyte/source-clockify:latest
{"type": "LOG", "log": {"level": "WARN", "message": "\n We did not find the airbyte/source-clockify:latest image for this connector. This probably means this version has not yet been published to an accessible docker registry like DockerHub."}}

 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestDiscovery.test_backward_compatibility[inputs0] s           88% ████████▉ 
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestBasicRead.test_read[inputs0] ✓                             92% █████████▎{"type": "LOG", "log": {"level": "ERROR", "message": "Docker container was failed, code 1, error:\n{\"type\": \"TRACE\", \"trace\": {\"type\": \"ERROR\", \"emitted_at\": 1665919692094.1528, \"error\": {\"message\": \"Something went wrong in the connector. See the logs for more details.\", \"internal_message\": \"2 validation errors for ConfiguredAirbyteCatalog\\nstreams -> 0 -> sync_mode\\n  value is not a valid enumeration member; permitted: 'full_refresh', 'incremental' (type=type_error.enum; enum_values=[<SyncMode.full_refresh: 'full_refresh'>, <SyncMode.incremental: 'incremental'>])\\nstreams -> 0 -> destination_sync_mode\\n  value is not a valid enumeration member; permitted: 'append', 'overwrite', 'append_dedup' (type=type_error.enum; enum_values=[<DestinationSyncMode.append: 'append'>, <DestinationSyncMode.overwrite: 'overwrite'>, <DestinationSyncMode.append_dedup: 'append_dedup'>])\", \"stack_trace\": \"Traceback (most recent call last):\\n  File \\\"/airbyte/integration_code/main.py\\\", line 13, in <module>\\n    launch(source, sys.argv[1:])\\n  File \\\"/usr/local/lib/python3.9/site-packages/airbyte_cdk/entrypoint.py\\\", line 123, in launch\\n    for message in source_entrypoint.run(parsed_args):\\n  File \\\"/usr/local/lib/python3.9/site-packages/airbyte_cdk/entrypoint.py\\\", line 111, in run\\n    config_catalog = self.source.read_catalog(parsed_args.catalog)\\n  File \\\"/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/source.py\\\", line 90, in read_catalog\\n    return ConfiguredAirbyteCatalog.parse_obj(self.read_config(catalog_path))\\n  File \\\"pydantic/main.py\\\", line 521, in pydantic.main.BaseModel.parse_obj\\n  File \\\"pydantic/main.py\\\", line 341, in pydantic.main.BaseModel.__init__\\npydantic.error_wrappers.ValidationError: 2 validation errors for ConfiguredAirbyteCatalog\\nstreams -> 0 -> sync_mode\\n  value is not a valid enumeration member; permitted: 'full_refresh', 'incremental' (type=type_error.enum; enum_values=[<SyncMode.full_refresh: 'full_refresh'>, <SyncMode.incremental: 'incremental'>])\\nstreams -> 0 -> destination_sync_mode\\n  value is not a valid enumeration member; permitted: 'append', 'overwrite', 'append_dedup' (type=type_error.enum; enum_values=[<DestinationSyncMode.append: 'append'>, <DestinationSyncMode.overwrite: 'overwrite'>, <DestinationSyncMode.append_dedup: 'append_dedup'>])\\n\", \"failure_type\": \"system_error\"}}}\n"}}

 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestBasicRead.test_airbyte_trace_message_on_failure[inputs0] ✓ 96% █████████▋
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_full_refresh.py::TestFullRefresh.test_sequential_reads[inputs0] ✓      100% ██████████
{"type": "LOG", "log": {"level": "INFO", "message": "/root/Development/airbyte/airbyte-integrations/connectors/source-clockify - SAT run - 74f7501f12c1b63ded1b2e3382e0779a48b9fd23 - PASSED"}}

========================================================================= short test summary info =========================================================================
SKIPPED [1] ../../bases/source-acceptance-test/source_acceptance_test/plugin.py:60: Skipping TestIncremental.test_two_sequential_reads because not found in the config
SKIPPED [1] ../../bases/source-acceptance-test/source_acceptance_test/tests/test_core.py:51: The previous connector image could not be retrieved.
SKIPPED [1] ../../bases/source-acceptance-test/source_acceptance_test/tests/test_core.py:229: The previous connector image could not be retrieved.

Results (51.70s):
      24 passed
       2 skipped

</details>

### CODE REVIEW

**Comment by marcosmarxm** [comment]
Can you provide the output of integration tests @RobLucchi ?

File: airbyte-integrations/connectors/source-clockify/source_clockify/spec.json
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-clockify/source_clockify/spec.json:16]:
```suggestion
            "api_key": {
```

File: airbyte-integrations/connectors/source-clockify/source_clockify/spec.json
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-clockify/source_clockify/spec.json:12]:
```suggestion
            "workspace_id": {
```

File: airbyte-integrations/connectors/source-clockify/source_clockify/streams.py
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-clockify/source_clockify/streams.py:20]:
```suggestion
        self.workspace_id = workspaceId

```

Please check Python PEP-8 recommendations for variable naming.

File: airbyte-integrations/connectors/source-clockify/source_clockify/streams.py
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-clockify/source_clockify/streams.py:48]: Please enable cache for this stream, so when reading the `TimeEntrie`won't read a second time.

File: airbyte-integrations/connectors/source-clockify/source_clockify/streams.py
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-clockify/source_clockify/streams.py:53]: Same here: enable cache.

**Review by marcosmarxm — CHANGES_REQUESTED** [review]
Some comments, please run the formatter command.

**Comment by sajarin** [comment]
@RobLucchi Thanks for the new source connector. This PR is part of AIrbyte community maintainer program and will be reviewed by a member of the community. Please be patient as you wait for a maintainer to be assigned to this PR. 

Thanks!

**Comment by RobLucchi** [comment]
@sajarin Requested Changes Applied, waiting for any further reviews.

**Comment by natalyjazzviolin** [comment]
Hey @RobLucchi ! I tested your connector locally and made a few changes to make sure the integration tests pass.

Could you please check over your schemas and either:
- Remove fields that are type `null`
- OR add a second type, so null is the fallback: `["null", "string"]` for example.

Looking through the API documentation, it looks like `costRate` (one of the null fields) can either be null, a string, or an object:
https://clockify.me/developers-api

**Comment by RobLucchi** [comment]
@natalyjazzviolin Could you please check the last changes.

**Comment by natalyjazzviolin** [comment]
@RobLucchi testing! Please note that the task duration field needs a fallback value of `null`! And another small detail - we would really appreciate it if you used a more descriptive branch name in the future :) something like `roblucchi/source-clockify` would work great!

What a cool source, thanks for the contribution!

**Comment by marcosmarxm** [comment]
@natalyjazzviolin before publishing please:
- [ ] add seed uuid into airbyte-config/.../source_seed file
- [ ] format files there are some EOF missing in the review
- [ ] publish :)

**Review by marcosmarxm — APPROVED** [review]

File: airbyte-config/init/src/main/resources/seed/source_definitions.yaml
Comment (marcosmarxm) [inline airbyte-config/init/src/main/resources/seed/source_definitions.yaml:197]: @natalyjazzviolin please remove this.

**Review by marcosmarxm — CHANGES_REQUESTED** [review]

**Review by marcosmarxm — CHANGES_REQUESTED** [review]

**Comment by marcosmarxm** [comment]
Dockerhub has an issue with push command. I'll wait until this is solved.

**Review by marcosmarxm — APPROVED** [review]

**Pull request merged**
